pkg/controller/controller_utils.go

```diff
@@ -549,6 +550,13 @@ func GetPodFromTemplate(template *v1.PodTemplateSpec, parentObject runtime.Objec
 	}
 	prefix := getPodsPrefix(accessor.GetName())
 
+	// if it is indexed job, add index information to its generateName prefix
+	if template.Annotations != nil {
+		if index, ok := template.Annotations[batch.JobCompletionIndexAnnotationAlpha]; ok {
```

Comment:
it's strange to add job-specific logic to a generic controller helper method. the other place that tweaks pod names for an index (statefulset) does so with its own method:

https://github.com/kubernetes/kubernetes/blob/a5cf298a95b4a60789a6ba1caab931bdc980aa2d/pkg/controller/statefulset/stateful_set_utils.go#L249-L256

Reply:
letting the CreatePodsWithControllerRef caller indicate a pod name suffix might make more sense than baking job-specific logic into this method

Reply:
`CreatePodsWithControllerRef` call `CreatePods`, and the function `createPods` will record an event which contains the created pod name. I think it’s better to add index suffix before outputting the event.

Or if we don’t use the generic method `CreatePods`, we can write something like statefulset, but that may need much effort?

Reply:
> letting the CreatePodsWithControllerRef caller indicate a pod name suffix might make more sense

I agree, but if a lot of callers have to change and the owners of the file are fine with this change as is, I think it's acceptable.

Just in case, @AliceZhang2016 could you check if adding a `CreatePodsWithControllerRefAndPrefix` to this file would be a lot of effort. I was kind of thinking in that direction when I created this cleanup #101156